fdbserver/Restore.actor.cpp

```diff
@@ -18,86 +18,604 @@
+}
+
+
+// Restore Worker
+ACTOR Future<Void> commitRestoreRoleInterfaces(Reference<RestoreWorkerData> self, Database cx) {
+	state ReadYourWritesTransaction tr(cx);
+	// For now, we assume only one role per restore worker
+	ASSERT( !(self->loaderInterf.present() && self->applierInterf.present()) );
+
+	loop {
+		try {
+			tr.reset();
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			ASSERT( !(self->loaderInterf.present() && self->applierInterf.present()) ); 
+			if ( self->loaderInterf.present() ) {
+				tr.set( restoreLoaderKeyFor(self->loaderInterf.get().id()), restoreLoaderInterfaceValue(self->loaderInterf.get()) );
+			}
+			if ( self->applierInterf.present() ) {
+				tr.set( restoreApplierKeyFor(self->applierInterf.get().id()), restoreApplierInterfaceValue(self->applierInterf.get()) );
 			}
+			wait (tr.commit() );
+			break;
+		} catch( Error &e ) {
+			printf("[WARNING]%s: commitRestoreRoleInterfaces transaction error:%s\n", self->describeNode().c_str(), e.what());
+			wait( tr.onError(e) );
 		}
 	}
 
-	//we are the leader
-	wait( delay(5.0) );
+	return Void();
+} 
+
+// Restore Worker
+ACTOR Future<Void> handleRecruitRoleRequest(RestoreRecruitRoleRequest req, Reference<RestoreWorkerData> self, ActorCollection *actors, Database cx) {
+	printf("[INFO][Worker] Node:%s get role %s\n", self->describeNode().c_str(),
+			getRoleStr(req.role).c_str());
+
+	while (self->isInProgress(RestoreCommandEnum::Recruit_Role_On_Worker)) {
+		printf("[DEBUG] NODE:%s handleRecruitRoleRequest wait for 1s\n",  self->describeNode().c_str());
+		wait(delay(1.0));
+	}
+	if ( self->isCmdProcessed(req.cmdID) ) {
+		req.reply.send(RestoreCommonReply(self->id(), req.cmdID));
+		return Void();
+	}
+	self->setInProgressFlag(RestoreCommandEnum::Recruit_Role_On_Worker);
 
-	state vector<RestoreInterface> agents;
+	if (req.role == RestoreRole::Loader) {
+		ASSERT( !self->loaderInterf.present() );
+		self->loaderInterf = RestoreLoaderInterface();
+		self->loaderInterf.get().initEndpoints();
+		self->loaderData = Reference<RestoreLoaderData>( new RestoreLoaderData(self->loaderInterf.get().id(),  req.nodeIndex) );
+		actors->add( restoreLoaderCore(self->loaderData, self->loaderInterf.get(), cx) );
+	} else if (req.role == RestoreRole::Applier) {
+		ASSERT( !self->applierInterf.present() );
+		self->applierInterf = RestoreApplierInterface();
+		self->applierInterf.get().initEndpoints();
+		self->applierData = Reference<RestoreApplierData>( new RestoreApplierData(self->applierInterf.get().id(), req.nodeIndex) );
+		actors->add( restoreApplierCore(self->applierData, self->applierInterf.get(), cx) );
+	} else {
+		TraceEvent(SevError, "FastRestore").detail("HandleRecruitRoleRequest", "UnknownRole"); //.detail("Request", req.printable());
+	}
+
+	wait( commitRestoreRoleInterfaces(self, cx) ); // Commit the interface after the interface is ready to accept requests
```

Comment:
Why not just reply with the interface instead of writing it to the database?

Reply:
The very first code uses DB to share the interfaces among workers. 
Both the restoreMaster and all restoreLoaders need the interface of all restoreAppliers.
Using DB for them to share the interfaces can simplify the code logic, I guess.